### PR TITLE
Replace RunTests scripts with .NET app

### PR DESCRIPTION
**Pull request by BrennanConroy**
Removing all our scripting logic from .cmd and .sh files for running tests on Helix and normalizing it into .NET code.

### CODE REVIEW

File: eng/helix/content/RunTests/Program.cs [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+                await ProcessUtil.RunAsync($"{dotnetRoot}/dotnet",
+                    "nuget add source https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet5/nuget/v3/index.json --configfile NuGet.config",
+                    environmentVariables: environmentVariables);
+
+                await ProcessUtil.RunAsync($"{dotnetRoot}/dotnet",
+                    "nuget list source",
+                    environmentVariables: environmentVariables);
+
+                await ProcessUtil.RunAsync($"{dotnetRoot}/dotnet",
+                    $"tool install dotnet-ef --global --version {efVersion}",
```
Comment (davidfowl) [inline eng/helix/content/RunTests/Program.cs]: Global?

File: eng/helix/content/RunTests/Program.cs [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+            {
+                Console.WriteLine(Path.GetFileName(file));
+            }
+            Console.WriteLine();
+
+            var discoveryResult = await ProcessUtil.RunAsync($"{dotnetRoot}/dotnet",
+                $"vstest {target} -lt",
+                environmentVariables: environmentVariables);
+
+            if (discoveryResult.StandardOutput.Contains("Exception thrown"))
```
Comment (davidfowl) [inline eng/helix/content/RunTests/Program.cs]: Why would this happen?

Comment (davidfowl) [reply eng/helix/content/RunTests/Program.cs]: Is it ever in stderr?

File: eng/helix/content/runtests.cmd [hunk 3]
```diff
@@ -24,89 +24,18 @@ powershell.exe -NoProfile -ExecutionPolicy unrestricted -Command "[Net.ServicePo
...
-
-%DOTNET_ROOT%\dotnet vstest %$target% -lt >discovered.txt
-find /c "Exception thrown" discovered.txt
-REM "ERRORLEVEL is not %ERRORLEVEL%" https://blogs.msdn.microsoft.com/oldnewthing/20080926-00/?p=20743/
-if not errorlevel 1 (
-    echo Exception thrown during test discovery. 1>&2
-    type discovered.txt 1>&2
-    exit /b 1
-)
+set ASPNETCORE_TEST_TARGET=%$target%
```
Comment (davidfowl) [inline eng/helix/content/runtests.cmd]: Any reason these aren't arguments?

File: eng/helix/content/RunTests/Program.cs (hunk 1 above)
Comment (BrennanConroy) [reply eng/helix/content/RunTests/Program.cs]: We set the `DOTNET_CLI_HOME` environment variable so it isn't "global". But maybe I should move some more of that logic out of the scripts

Comment (davidfowl) [reply eng/helix/content/RunTests/Program.cs]: Yes

File: eng/helix/content/RunTests/Program.cs (hunk 2 above)
Comment (BrennanConroy) [reply eng/helix/content/RunTests/Program.cs]: I'm not familiar with the specifics of this
https://github.com/dotnet/aspnetcore/pull/8107

File: eng/helix/content/RunTests/Program.cs [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+            else
+            {
+                Console.WriteLine("Running non-quarantined tests.");
+
+                // Filter syntax: https://github.com/Microsoft/vstest-docs/blob/master/docs/filter.md
+                var result = await ProcessUtil.RunAsync($"{dotnetRoot}/dotnet",
+                    commonTestArgs + " --TestCaseFilter:\"Quarantined!=true\"",
+                    environmentVariables: environmentVariables,
+                    outputDataReceived: Console.WriteLine,
+                    errorDataReceived: Console.WriteLine,
```
Comment (davidfowl) [inline eng/helix/content/RunTests/Program.cs]: Console.Error.WriteLine

File: eng/helix/content/RunTests/Program.cs [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+namespace RunTests
+{
+    class Program
+    {
+        static async Task Main(string[] args)
+        {
+            if (args.Length < 7)
+            {
+                Console.WriteLine($"Expected at least 7 args, got {args.Length}.");
```
Comment (davidfowl) [inline eng/helix/content/RunTests/Program.cs]: 😄

File: eng/helix/content/RunTests/ProcessUtil.cs [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+                    Arguments = arguments,
+                    RedirectStandardOutput = true,
+                    RedirectStandardError = true,
+                    UseShellExecute = false,
+                    CreateNoWindow = true,
+                },
+                EnableRaisingEvents = true
+            };
+
+
```
Comment (davidfowl) [inline eng/helix/content/RunTests/ProcessUtil.cs]: Can we Console.WriteLine the command being executed in here?

**Comment by BrennanConroy** [comment]
Looks like I've got all tests running correctly now. So just need to do cleanup.

**Comment by dougbu** [comment]
I was invited to review this but have no context. What are the aims of this PR?

**Comment by BrennanConroy** [comment]
Just figured you might be interested.

> What are the aims of this PR?

Replacing our scripts with .NET code. We know how to write in .NET and it consolidates our logic to one place instead of multiple.

**Comment by davidfowl** [comment]
Ship it

**Comment by dougbu** [comment]
> Just figured you might be interested.

The overall approach is not a big deal to me either way.

I am concerned about requiring that Helix agents build the new RunTests project. That will likely slow down every run enough to matter cumulatively (we do lots 'o runs). Suggest building _then_ including the results in the job payloads.

> We know how to write in .NET and it consolidates our logic to one place instead of multiple.

nit: Not sure how valid the first point is given everyone needs to be familiar w/ MSBuild and scripting to do ops-on-call work. But, I agree maintaining `bash` and `cmd` scripts in parallel can get frustrating.

**Review by davidfowl — APPROVED** [review]

**Review by HaoK — APPROVED** [review]
Nice, this is way better than maintaining cmd and sh versions!

**Comment by dougbu** [comment]
@BrennanConroy unless you've confirmed the time spent on Helix agents hasn't worsened, please file a follow-up issue to move the RunTests.csproj build into the main job.

**Comment by BrennanConroy** [comment]
It takes a couple seconds to build... this is definitely not an area we need to care about to improve PR/rolling builds...

**Comment by HaoK** [comment]
Well I guess its probably a little cleaner just to build the runtests thing as part of the normal builds, and just include the dll into helix content and run that dll in the scripts, since that would let us write unit tests against RunTests too, which probably is a good thing since its gotten pretty complicated now :/

**Pull request merged**